### PR TITLE
Simplifications & general improvements

### PR DESCRIPTION
**Pull request by ihrwein**
Changes
* Operation and FutureOperation has been removed
* Considerable parts of the README was moved to the crate's doc
* source code was reformatted with `cargo fmt`
* breaking changes documented

### CODE REVIEW

**Comment by ihrwein** [comment]
@coolreader18 may I ask you to review the PR?

File: src/retry/async.rs [hunk 1]
```diff
@@ -256,17 +227,17 @@ async fn f() -> Result<(), backoff::Error<&'static str>> {
...
 ```"#),
                 pub fn retry<I, E, Fn, B>(
                     backoff: B,
                     operation: Fn,
-                ) -> Retry<impl Sleeper, B, NoopNotify, Fn, Fn::Fut>
+                ) -> Retry<impl Sleeper, B, NoopNotify, Fn, I, E>
                 where
                     B: Backoff,
-                    Fn: FutureOperation<I, E>,
+                    Fn: Future<Output = Result<I, Error<E>>>,
```
Comment (coolreader18) [inline src/retry/async.rs:240]: I think you have to have another generic `Fut` here, and then `Fn: FnMut() -> Fut`

**Review by coolreader18 — APPROVED** [review]
Looks good overall, just the thing about `Future` vs `FnMut() -> Future`. I'm pretty sure with how it is currently, retry would panic once it attempts the second try, since the Future from the `async {}` block would be exhausted.

**Comment by ihrwein** [comment]
Thank you for the review!

**Review by coolreader18 — APPROVED** [review]

**Pull request merged**